### PR TITLE
Adding RetentionMillis at kafka channel spec level

### LINKED ISSUE #643: Ability to provide Topic Retention Period on Kafka Channel Spec
**Problem**
Currently, kafka Channel spec does not allow passing in Retention period at the CRD level.  

The distributed Channel already has a util function to get Retention Periood and uses the default value to create Topic using  admin client. A code change would be required here - https://github.com/knative-sandbox/eventing-kafka/blob/fbb9090ab2a9407cdcbc562d54429c20518930db/pkg/channel/distributed/controller/util/channel.go#L106

For consolidated channel, we need to pass in the retention period under Config Entries like done here -https://github.com/knative-sandbox/eventing-kafka/blob/fbb9090ab2a9407cdcbc562d54429c20518930db/pkg/channel/distributed/controller/kafkachannel/topic.go#L97 while creating topic here - https://github.com/knative-sandbox/eventing-kafka/blob/4027f852ca6ca0ccdad0a062eec7c531d82a9591/pkg/channel/consolidated/reconciler/controller/kafkachannel.go#L544

Other changes to accept the new spec would also be needed.

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)**
A user using Kafka Channel and configuring Retention Period at topic level

**Exit Criteria**
Kafka Channel retention period is configurable at the CR level.

### PR DESCRIPTION
**Pull request by ssandanshi**
Fixes #643

## Proposed Changes

- 🎁  Kafka Channel RetentionMillis can now be provided on Channel spec. Works for both Consolidated and Distributed Channel. Its an Optional field which if absent will default to the defaultRetentionMillis value provided in the Common ConfigMap. If the defaultRetentionMillis is not provided in config map, the default value of 604800000(1 Week) is taken from constants in code.

**Release Note**

```release-note
Kafka Channel RetentionMillis can now be provided on Channel spec. Works for both Consolidated and Distributed Channel. Its an Optional field which if absent will default to the value provided in the Common ConfigMap
```

### CODE REVIEW

**Comment by ssandanshi** [comment]
My first PR here so any comments/suggestions welcome. I did manually test the behavior for consolidated and distributed channels.

Let me know if I'm missing any tests somewhere or anything else.

CC: @slinkydeveloper @eric-sap (since you worked on the configmap consolidation)

**Comment by ssandanshi** [comment]
@googlebot I signed it!

File: config/channel/resources/kafkachannel-crd.yaml [hunk 1]
```diff
@@ -48,6 +48,9 @@ spec:
                   type: integer
                   maximum: 32767
                   default: 1
+                retentionMillis:
+                  description: retentionMillis is the topic retention period (in ms) of a Kafka Topic. This is an optional field. By default, the Channel default retention time would be used
+                  type: integer
```
Comment (travis-minke-sap) [inline config/channel/resources/kafkachannel-crd.yaml:53]: Might want to say that the "default" Channel retention is configured in the ConfigMap ?

File: pkg/apis/messaging/v1beta1/kafka_channel_types.go [hunk 2]
```diff
@@ -67,6 +67,10 @@ type KafkaChannelSpec struct {
 	// ReplicationFactor is the replication factor of a Kafka topic. By default, it is set to 1.
 	ReplicationFactor int16 `json:"replicationFactor"`
 
+	// RetentionMillis is the retention Time period (in millisecond) of a Kafka topic.
+	// +optional
+	RetentionMillis int64 `json:"retentionMillis,omitempty"`
+
```
Comment (travis-minke-sap) [inline pkg/apis/messaging/v1beta1/kafka_channel_types.go:73]: Did you re-run `update-codegen.sh` ?    Should there be other additions like validate not less than 0 in `kafka_channel_validation.go` ?

**Comment by travis-minke-sap** [comment]
We originally had support for `retentionMillis` in the original version of the Distributed KafkaChannel but lost it while aligning with the Consolidated implementation - it's been on my list to add it back for a long time now - thanks for actually getting it done!

File: config/channel/resources/kafkachannel-crd.yaml (hunk 1 above)
Comment (ssandanshi) [reply config/channel/resources/kafkachannel-crd.yaml:53]: Cleared up the language here.

File: config/channel/resources/kafkachannel-crd.yaml [hunk 3]
```diff
@@ -48,6 +48,9 @@ spec:
                   type: integer
                   maximum: 32767
                   default: 1
+                retentionMillis:
+                  description: retentionMillis is the topic retention period (in ms) of a Kafka Topic. This is an optional field. By default, the Channel default retention time provided in the ConfigMap would be used. If the default value is not provided in Config Map, 604800000(1 week) is used.
```
Comment (slinkydeveloper) [inline config/channel/resources/kafkachannel-crd.yaml:52]: Should we use the ISO8601 format instead? This is what we use in other eventing apis, for example in the DeliverySpec https://github.com/knative/eventing/blob/main/pkg/apis/duck/v1/delivery_types.go#L52

**Comment by travis-minke-sap** [comment]
Closing now that PR #828 has been merged as part of Issue #817.

**Pull request closed without merging**